File: 9/997/FindTheTownJudge.py
# https://leetcode.com/problems/find-the-town-judge/description/?envType=daily-question&envId=2024-02-22

class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:

        def binarySearch(arrl, val):
            arrl.sort()
            left=0
            right=len(arrl)-1
            while(left<=right):
                mid=left+(right-left)//2
                if arrl[mid]==val:
                    return True
                elif arrl[mid]>val:
                    right=mid-1
                else:
                    left=mid+1
            return False

        def find(arrl, val):
            res=binarySearch(arrl, val)
            return res

        sz=len(trust)
        if sz==0:
            if n!=1:
                return -1
            return 1
        pplsmp={}
        for i in range(0, sz, 1):
            if trust[i][0] in pplsmp:
                vec=pplsmp[trust[i][0]]
                vec.append(trust[i][1])
                pplsmp[trust[i][0]]=vec
            else:
                pplsmp[trust[i][0]]=[]
                pplsmp[trust[i][0]].append(trust[i][1])
        for i in range(1, n+1, 1):
            if i not in pplsmp:
                flag1=False
                for j in range(1, n+1, 1):
                    if j!=i:
                        if j in pplsmp:
                            arrl=pplsmp[j]
                            flag2=find(arrl, i)
                            if(flag2==False):
                                flag1=True
                                break
                        else:
                            flag1=True
                            break
                if flag1==False:
                    return i
        return -1

File: 9/997/FindTheTownJudge.cpp
/*
https://leetcode.com/problems/find-the-town-judge/description/?envType=daily-question&envId=2024-02-22
*/

class Solution {
public:
    int findJudge(int n, vector<vector<int>>& trust) {
        int sz=trust.size();
        if(sz==0){
            if(n!=1) return -1;
            return 1;
        }
        map<int, vector<int>>pplsmp;
        for(auto &trustentries: trust){
            pplsmp[trustentries[0]].push_back(trustentries[1]);
        }
        for(int i=1; i<=n; ++i){
            auto itr=pplsmp.find(i);
            if(itr==pplsmp.end()){
                bool flag=false;
                for(int j=1; j<=n; ++j){
                    if(j!=i){
                        auto itr2=pplsmp.find(j);
                        if(itr2!=pplsmp.end()){
                            auto vec=pplsmp[j];
                            auto itr2=find(vec.begin(), vec.end(), i);
                            if(itr2==vec.end()){
                                flag=true;
                                break;
                            }
                        } else{
                            flag=true;
                            break;
                        }
                    }
                }
                if(!flag) return i;
            }
        }
        return -1;
    }
};